### PR TITLE
NEW: Probe variation with coherent probe modes

### PR DESCRIPTION
**Pull request by carterbox**
## Purpose

Add a model for probe variation over time with coherent probe modes to the least-squares algorithm and fix a problem with phase ramping at the edges of the reconstruction.

## Approach

Please read new documentation in `tike.ptycho.probe` for information about the variation correction.

To fix the phase ramping problem, I re-implemented the update directions in the least squares method to use weighted sums instead of normalization by division. The one side effect is that the least-squares method seems to converge more slowly now. I will have to see if momentum acceleration can improve convergence rates.

Forced orthogonalization for incoherent probes is now disabled in the least-squares method because it was causing artifacts in my tests.

### Submitter
- [x] Write a helpfully descriptive pull request title.
- [x] Organize changes into logically grouped commits with descriptive commit messages.
- [x] Document all new functions.
- [x] Write tests for new functions or explain why they are not needed.
- [ ] Build the documentation successfully
- [x] Use [`yapf`](https://github.com/google/yapf) to format python code.

### Reviewer
- [ ] Actually read all of the code.
- [ ] Run the new code yourself.
- [ ] Write a summary of the changes as you understand them.
- [ ] Thank the submitter.

### CODE REVIEW

File: src/tike/opt.py [hunk 1]
```diff
@@ -16,6 +16,83 @@
 randomizer = np.random.default_rng()
 
 
+def adagrad(d, v=None, eps=1e-6):
```
Comment (xiaodong-yu) [inline src/tike/opt.py:19]: What is this function for? I don't see any utilization of it.

File: tests/test_ptycho.py [hunk 2]
```diff
@@ -296,11 +307,59 @@ def test_consistent_lstsq_grad(self):
             },
         )
 
+    def test_consistent_lstsq_grad_variable_probe(self):
```
Comment (xiaodong-yu) [inline tests/test_ptycho.py:310]: Does the probe variation only work with ls gradient?

File: src/tike/opt.py (hunk 1 above)
Comment (carterbox) [reply src/tike/opt.py]: Sorry! I left in some commits from unrelated work.

File: tests/test_ptycho.py (hunk 2 above)
Comment (carterbox) [reply tests/test_ptycho.py:310]: Yes. Adding this feature to `cgrad` will require some thought and some refactoring, so I have not included it here.

**Review by xiaodong-yu — APPROVED** [review]

**Pull request merged**